### DISCUSSION TITLE
Getting "512 worker_connections are not enough" on High Load

### DISCUSSION THREAD

**Discussion by maieredv-manuel** [ZoeyVid/NPMplus#1316] [Q&A] [answered]
Hey! 

were using NPMPlus for 23 Proxy and 11 Stream Hosts.
Now we have the problem when our internal users are accessing the internal applications we are getting the "512 worker_connections are not enough" Message.

I found that there is an setting in the ngnix.conf file, but i think when ive edit this file, on the next update it will be changed to default!? Also there is an option that nginx can handle the worker_connection itselfs.

Can he have an ENV option for Settings? Or any otherway?

**Comment by Zoey2936** [comment]
>  Also there is an option that nginx can handle the worker_connection itselfs.

which exactly?

> Can he have an ENV option for Settings? Or any otherway?

if nginx can do this automatic then I would do this, if not then it depends if increasing this value has any downsides, if not then I can jsut increase the default

**Comment by maieredv-manuel** [comment]
https://www.sabatino.dev/misconfigured-nginx-worker_connections-caused-downtime/ 

I went to the nginx docs to read up about this parameter, and by default this was configured to 512.

Because of our many simultaneous websocket connections, we were effectively hitting the 512 limit, and NGINX was not configured to increase the worker_processes, so when our application server would hit this limit, it stopped accepting any new incoming requests.

I resolved this by increasing our [worker_connections](https://nginx.org/en/docs/ngx_core_module.html?ref=sabatino.dev#worker_connections) to 768 and I changed the [worker_processes](https://nginx.org/en/docs/ngx_core_module.html?ref=sabatino.dev#worker_processes) configuration from 1 to auto, so nginx would add extra workers if needed.

```
worker_processes auto;
events {
	worker_connections 768;
}
```

**Comment by Zoey2936** [comment]
the default for one of the values should already be auto:
https://github.com/ZoeyVid/NPMplus/blob/e5ecf35a89d6b18e4ab50a2199383765408672ba/rootfs/usr/local/nginx/conf/nginx.conf#L5

**Comment by maieredv-manuel** [comment]
Ah okay, thats the setting for allocating all CPU Cores - thats good.
Now we need the "worker_connections", i didnt found that setting in the nginx.conf !? that means when nothing is set in the config file, the default is set!?

**Comment by Zoey2936** [comment]
It should use the value written in nginx docs

**Comment by maieredv-manuel** [comment]
So can we change the value via "/data/nginx_custom/events.conf" or gets that file also an "reset" when updating?

**Comment by maieredv-manuel** [comment, marked as answer]
Ive changed the "/data/nginx_custom/events.conf" with "worker_connections 768;" and restartet the container, actually it started, but i dont know if that really changed anything ... do you know a way to check the current loaded config or get the actual worker_connections?

**Comment by Zoey2936** [comment]
`docker exec npmplus nginx -T` should output the full loaded config

**Comment by maieredv-manuel** [comment]
thats saadly only the output of all configs loaded, so only the "include /data/nginx_custom/events.conf;" is in the output, not if the process really uses the changed "worker_connections" value :-/

But okay, ill test it after the holidays, when next time all users are in the office :-) ill report again! 

Thanks for your help and merry christmas!

**Comment by Zoey2936** [comment]
Do you think increasing worker_connections by default is a bad idea?
And how many cpu cores has this system? Since worker_procces should be identical to the number of cpu threads.

Any many thanks!

**Comment by maieredv-manuel** [comment]
Puuuh, i dont know if its needed for all installations. 
When it can be done by editing that one file, its not that big deal.

My machine has only 1 CPU Core, but that wasnt the problem, cpu load is at ~5-10% at this time when ive got the "512 worker_connections are not enough" errors. 
I think auto is a good setting for "worker_processes" so nginx can scale up when its needed.

**Comment by Zoey2936** [comment]
The total connections are calculated by processes*connection, so doubling the cpu cores would double the total connections

**Reply by pagefault** [reply]
We're having the same problem (see #1685). Our VM has 4 cores and at the point where nginx runs out of connections, the CPU load is only at 40%. So we're having downtimes even though the server isn't even at half load. So it would be great to have a setting to increase this.

**Reply by maieredv-manuel** [reply]
Take a look at the compose.yaml Template, there is an option to change the amount of worker prozesses:
- "NGINX_WORKER_PROCESSES=8" value of worker_processes, default and recommended: auto

**Reply by pagefault** [reply]
thanks. Should I set it to 8 or to auto? Should I also increase connetions? I couldn't find the file you were talking about. /data/nginx_custom/events.conf looks like an empty file for me.

**Reply by maieredv-manuel** [reply]
Its this one: https://github.com/ZoeyVid/NPMplus/blob/develop/compose.yaml here are all Variables listed.
Provided by Zoey. 

You need to check whats best for your Installation. 

When i had the Problem the option for controlling the worker processes wasnt there, i think Zoey changed the Locations of the files, because its not needed anymore ... or ive changed the File in the container directly ... i dont know it anymore ;-) 

But controlling the worker processes directly is the safer way - and you can do it via compose.yaml 👍

**Reply by pagefault** [reply]
Thanks for the reply. I would like it to handle at least 10,000 connections. If we are stuck with the 512 connetions limit per process, I would need to run about 20 processes. That doesn't sound very smart. I'd love it if I could increase the connections at least to 1024 per process. These default values are from 15 years ago..

**Reply by maieredv-manuel** [reply]
Yeah thats true ... ive looked again in my installation ... when i connect directly in to my container, the file is there, i dont know if the file was there or ive created it ... i only have that single Line in it "worker_connections 768;" so i think its not a default file. 

The "default" nginx.conf (https://github.com/ZoeyVid/NPMplus/blob/develop/rootfs/usr/local/nginx/conf/nginx.conf) file calls the "/data/custom_nginx/events.conf" so i think ive created it and its not a default file.

**Reply by maieredv-manuel** [reply]
ive looked futher ...
Its a file in the "Data-Volume-Bind" so its defintly custom.
So you only need to create that file, set the Value you want and it should work.

![image](https://github.com/user-attachments/assets/ebd613c2-88f6-4d7a-9f47-960b94c3a608)

**Reply by pagefault** [reply]
Yes you are correct, I wrote that line into the empty events.conf file and it seems to have worked after a restart. Thank you for the help.

**Reply by maieredv-manuel** [reply]
Nice to hear 👍

**Comment by maieredv-manuel** [comment]
Aha, okay ... 
That makes sense - but the CPU load isnt the problem, like i said, we have around 10% Load at this time ...

But i think for the most installations with more then one CPU Core it shouldt be an problem.

**Reply by Zoey2936** [reply]
Yes thats true, but the cpu cores definde the number of total processes

**Reply by maieredv-manuel** [reply]
Sure, but when we can adjust the worker_processes with the "worker_connections" Setting in "/data/nginx_custom/events.conf" its not a big deal to do that :-) Why should i change my Cloud VM to something bigger with more costs when i can change the behave of nginx with a simple setting? :-)

I think the Setting in that file is also peristent for updates because its an file in the "Data" Folder that is accessible for all, so there is also no need for changing anything from your side.

**Reply by Zoey2936** [reply]
yes this file is persistent

**Reply by maieredv-manuel** [reply]
My Chef said we update the VM anyways :-D So all its all good, but also good to know that there is an option to change it, without the need of upgrading :-)

**Comment by Zoey2936** [comment]
I will convert this to a discussion
